src/s2e/Plugins/ExecutionMonitors/LibraryCallMonitor.cpp

```diff
@@ -1,145 +1,208 @@
 #include <s2e/Plugins/OSMonitors/OSMonitor.h>
+#include <s2e/Plugins/OSMonitors/Support/ModuleMap.h>
+#include <s2e/Plugins/OSMonitors/Support/ProcessExecutionDetector.h>
 #include <s2e/S2E.h>
 #include <s2e/Utils.h>
-#include <s2e/cpu.h>
 
-#include <iostream>
+#include <list>
+#include <unordered_map>
 
 #include "LibraryCallMonitor.h"
 
 namespace s2e {
 namespace plugins {
 
-S2E_DEFINE_PLUGIN(LibraryCallMonitor, "Flags all calls to external libraries", "LibraryCallMonitor", "OSMonitor",
-                  "FunctionMonitor", "ModuleExecutionDetector", "Vmi");
+S2E_DEFINE_PLUGIN(LibraryCallMonitor, "Monitors external library function calls", "", "ModuleMap", "OSMonitor",
+                  "ProcessExecutionDetector", "Vmi");
 
-void LibraryCallMonitor::initialize() {
-    m_detector = s2e()->getPlugin<ModuleExecutionDetector>();
-    m_functionMonitor = s2e()->getPlugin<FunctionMonitor>();
-    m_vmi = s2e()->getPlugin<Vmi>();
-    m_monitor = static_cast<OSMonitor *>(s2e()->getPlugin("OSMonitor"));
+/////////////////////////////
+// LibraryCallMonitorState //
+/////////////////////////////
 
-    ConfigFile *cfg = s2e()->getConfig();
-    m_displayOnce = cfg->getBool(getConfigKey() + ".displayOnce", false);
+///
+/// The \c LibraryCallMonitor state is essentially a LRU cache. This is so that we don't need to search through all the
+/// loaded modules for every external library call to find its target.
+///
+/// Adapted from https://github.com/lamerman/cpp-lru-cache
+///
+class LibraryCallMonitorState : public PluginState {
+public:
+    typedef std::pair<const ModuleDescriptor *, std::string> Export;
+
+private:
+    typedef std::pair<uint64_t, Export> AddressExportPair;
+    typedef std::list<AddressExportPair>::iterator Iterator;
 
-    bool ok = false;
+    std::list<AddressExportPair> m_cacheList;
+    std::unordered_map<uint64_t, Iterator> m_cacheMap;
 
-    // Fetch the list of modules where to report the calls
-    ConfigFile::string_list moduleList =
-        cfg->getStringList(getConfigKey() + ".moduleIds", ConfigFile::string_list(), &ok);
+    const static size_t MAX_CACHE_SIZE = 50;
 
-    if (!ok || moduleList.empty()) {
-        getWarningsStream() << "no modules specified, tracking everything.\n";
+public:
+    virtual LibraryCallMonitorState *clone() const {
+        return new LibraryCallMonitorState(*this);
     }
 
-    foreach2 (it, moduleList.begin(), moduleList.end()) {
-        if (!m_detector->isModuleConfigured(*it)) {
-            getWarningsStream() << "module " << *it << " is not configured\n";
-            exit(-1);
-        }
-        m_trackedModules.insert(*it);
+    static PluginState *factory(Plugin *plugin, S2EExecutionState *state) {
+        return new LibraryCallMonitorState();
     }
 
-    m_detector->onModuleLoad.connect(sigc::mem_fun(*this, &LibraryCallMonitor::onModuleLoad));
+    /// Add a new export to the cache at the given address
+    void put(uint64_t address, const Export &exp) {
```

Comment:
I presume many plugins would like to have this info.
Would it make more sense to add this to ModuleMap instead? And then initialize this info lazily for interested clients?